File: JS/Linked Lists/SentinelLinkedList.js
/*
 * --------------------------------------------------------------------------------
 * File :         SentinelLinkedList.js
 * Project :      JS
 * Author :       Saurish Phatak
 *
 *
 * Description : Sentinel Linked List in JavaScript
 * --------------------------------------------------------------------------------
 *
 * Revision History :
 * 2020-August-02	[SP] : Created
 * --------------------------------------------------------------------------------
 * Copyright (c) 2020 Saurish Phatak
 *
 * License :
 *
 *    This code is meant for learning.
 *    Do NOT Copy Paste it as it may not help in understanding the code.
 *    You are required to understand and type it yourself.
 *
 * DISCLAIMER :
 *    This code may contain unintentional and intentional bugs.
 *    There are no guarantees of this code working properly.
 * --------------------------------------------------------------------------------
 */

// Represents a value in the List
class ListNode {
    constructor(value) {
        this.value = value;
        this.previous = this.next = null;
    }
}

// Represents a Sentinel Linked List
class SentinelLinkedList {
    // Point to the head and tail dummy nodes
    #head;
    #tail;

    constructor() {
        // Create dummy head and tail nodes
        this.#head = { previous: null, value: undefined, next: null };
        this.#tail = { previous: null, value: undefined, next: null };

        // Initially head and tail point to each other
        this.#head.next = this.#tail;
        this.#tail.previous = this.#head;
    }

    // Adds a Node to the back of the List
    pushBack(value) {
        // Create a newNode
        let newNode = new ListNode(value);

        // Think of the List as a wall
        // Make the newNode climb up the wall
        newNode.next = this.#tail;
        newNode.previous = this.#tail.previous;

        // Now pull the newNode up the wall
        this.#tail.previous.next = newNode;
        this.#tail.previous = newNode;
    }

    // Removes a value from the List
    remove(value) {

        // Search for the value
        for (let current = this.#head.next; current != this.#tail; current = current.next) {
            // Value found
            if (value == current.value) {
                // Disconnect current Node from the List
                current.previous.next = current.next;
                current.next.previous = current.previous;

                return true;
            }
        }

        // Value not found in the List
        return false;
    }

    // Prints the List forward
    printForward() {
        let output = "";

        for (let current = this.#head.next; current != this.#tail; current = current.next) {
            output += current.value;

            if (current.next != this.#tail)
                output += "<->";
        }

        console.log(output);
    }

    // Prints the List backward
    printBackward() {
        let output = " ";

        for (let current = this.#tail.previous; current != this.#head; current = current[previous]) {
            output += current[value];

            if (current[previous] != this.#head)
                output += "<->";
        }

        console.log(output);
    }

    // Removes all the values from the List
    removeAll() {
        // Start from head.next and keep on disconnecting
        // Nodes from the List until tail is reached
        while (this.#head.next != this.#tail) {
            // Make head.next point to head.next.next
            this.#head.next = this.#head.next.next;

            // Disconnect the previous Node from the List
            this.#head.next.previous.next = null;
            this.#head.next.previous = null;
        }
    }
}

// Create a Sentinel Linked List
let list = new SentinelLinkedList();

list.pushBack(5);
list.pushBack(7);
list.pushBack(1);

list.printForward();

list.removeAll();
list.printForward();

list.remove(7);
list.printForward();

list.remove(1);
list.printForward();

list.remove(5);
list.printForward();

File: CPP/Linked Lists/SentinelLinkedList.cpp
/*
 * --------------------------------------------------------------------------------
 * File :         SentinelLinkedList.cpp
 * Project :      CPP
 * Author :       Saurish Phatak
 *
 *
 * Description : Sentinel Linked List in C++
 * --------------------------------------------------------------------------------
 *
 * Revision History :
 * 2020-August-01	[SP] : Created
 * --------------------------------------------------------------------------------
 */

#include <iostream>

using namespace std;

// Node represents a value in the List
struct Node
{
    // Points to the previous Node
    Node *previous;

    // Holds the Node's value
    int value;

    // Points to the next Node
    Node *next;

    // Constructor
    Node(int v)
    {
        previous = next = nullptr;
        (*this).value = v;
    }
};

// Represents the SentinelLinkedList
class SentinelLinkedList
{
    // Points to the Dummy Head
    Node *head;

    // Points to the Dummy Tail
    Node *tail;

public:
    // Constructor
    SentinelLinkedList()
    {
        // Create dummy Nodes
        head = tail = new Node(0);

        // Point head and tail to each other
        head->next = tail;
        (*tail).previous = head;
    }

    // Method to add a Node in the Linked List
    void pushBack(int value)
    {
        /**
         * This is where the Sentinel Linked List
         * really SHINES!
         * We don't have to check if this the First Node
         * that we're adding.
         *
         * We simply create a newNode and add it before the
         * tail.
         */

        // Holds the new Node
        Node *newNode;

        // If the allocation failed
        if (!(newNode = new Node(value)))
            return;

        /**
         * Add the newNode before the tail
         * Think of the List as a wall
         */
        // Make the newNode climb up the wall
        newNode->next = tail;
        newNode->previous = tail->previous;

        // Pull the newNode up the wall
        newNode->previous->next = newNode;
        tail->previous = newNode;

        // And that's it! That's how easy it is
        // to add a new Node to the Sentinel List
    }

    // Method to remove a Node from the List
    bool remove(int valueToRemove)
    {
        // Search for the value
        // Start from the head's next and not the head
        // as head itself points to a Dummy Node
        for (Node *current = head->next; current != tail; current = current->next)
        {
            // Value found
            if ((*current).value == valueToRemove)
            {
                // Disconnect the current Node from the List 
                current->previous->next = current->next;
                current->next->previous = current->previous;

                // Delete the current Node
                delete current;
                return true;
            }
        }

        // Value not in the List
        return false;
    }

    // Method to print the List in Forward direction
    void printForward()
    {
        // Start from head's next and go upto tail
        for (Node *current = head->next; current != tail; current = current->next)
        {
            // Print the current Node's value
            cout << current->value << " ";
        }

        cout << endl;
    }
};

int main()
{
    // Create a new Sentinel List
    SentinelLinkedList list;

    // Add some values to the list
    int value;
    while (cout << "Enter value (0 to stop) : ",
           cin >> value,
           value)
    {
        list.pushBack(value);
    }

    return 0;
}